SYCL/UserDefinedReductions/user_defined_reductions.cpp

```diff
@@ -0,0 +1,209 @@
+                                     lhs.ull + rhs.ull);
+}
+
+template <typename T, std::size_t... Is>
+constexpr std::array<T, sizeof...(Is)> init_array(T value,
+                                                  std::index_sequence<Is...>) {
+  return {{(static_cast<void>(Is), value)...}};
+}
+
+using namespace sycl;
+
+template <typename InputContainer, typename OutputContainer,
+          class BinaryOperation>
+void test(queue q, InputContainer input, OutputContainer output,
+          BinaryOperation binary_op, size_t workgroup_size,
+          typename OutputContainer::value_type identity,
+          typename OutputContainer::value_type init) {
+  using InputT = typename InputContainer::value_type;
+  using OutputT = typename OutputContainer::value_type;
+  constexpr size_t N = input.size();
+  {
+    buffer<InputT> in_buf(input.data(), input.size());
+    buffer<OutputT> out_buf(output.data(), output.size());
+
+    q.submit([&](handler &cgh) {
+      accessor in{in_buf, cgh, sycl::read_only};
+      accessor out{out_buf, cgh, sycl::write_only, sycl::no_init};
+
+      size_t temp_memory_size = workgroup_size * sizeof(InputT);
+      auto scratch = sycl::local_accessor<std::byte, 1>(temp_memory_size, cgh);
+      cgh.parallel_for(
+          nd_range<1>(workgroup_size, workgroup_size), [=](nd_item<1> it) {
+            // Create a handle that associates the group with an allocation it
+            // can use
+            auto handle =
+                sycl::ext::oneapi::experimental::group_with_scratchpad(
+                    it.get_group(), sycl::span(&scratch[0], temp_memory_size));
+
+            InputT *first = in.get_pointer();
+            InputT *last = first + N;
+            // check reduce_over_group w/o init
+            out[0] = sycl::ext::oneapi::experimental::reduce_over_group(
+                handle, in[it.get_global_id(0)], binary_op);
+
+            // check reduce_over_group with init
+            out[1] = sycl::ext::oneapi::experimental::reduce_over_group(
+                handle, in[it.get_global_id(0)], init, binary_op);
+
+            // check joint_reduce w/o init
+            out[2] = sycl::ext::oneapi::experimental::joint_reduce(
+                handle, first, last, binary_op);
+
+            // check joint_reduce with init
+            out[3] = sycl::ext::oneapi::experimental::joint_reduce(
+                handle, first, last, init, binary_op);
+          });
+    });
+    q.wait();
+  }
+  assert(output[0] == std::reduce(input.begin(), input.begin() + workgroup_size,
+                                  identity, binary_op));
+  assert(output[1] == std::reduce(input.begin(), input.begin() + workgroup_size,
+                                  init, binary_op));
+  assert(output[2] ==
+         std::reduce(input.begin(), input.end(), identity, binary_op));
+  assert(output[3] == std::reduce(input.begin(), input.end(), init, binary_op));
+}
+
+int main() {
```

Comment:
Please add a test with reduce over sub_group when one (or even several) WG(s) contain(s) several SGs. Might be done in another PR. The main thing I want is to use such a test as an example showing that even for reduce over SG the size of the temporary memory depends on WG size.